### PR TITLE
Event interface

### PR DESCRIPTION
**Pull request by bclothier**
This restores the implementation of events as an instance member of their containing class rather than as a static methods. This may potentially fix the issues where we miss some `VBComponents` events because only the first registration are considered. 

Note that this is somehow a reversion to an earlier pull request - see @comintern PRs here:
[PR to make it thread-safe](https://github.com/rubberduck-vba/Rubberduck/pull/2677)
[PR to plug memory leaks](https://github.com/rubberduck-vba/Rubberduck/pull/2687)

That said, there are still few differences --

1) We build upon the `SafeComWrapper`, creating `SafeEventedComWrapper` which controls the disposal.
2) We implement the source interface directly, rather than using a separate sink, which simplify the instance management as there's now an one-one relationship between the source and the events.

This is being put up for review by others before I proceed further on other and to help us decide if it's still warranted.

### CODE REVIEW

File: Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            lock (_lock)
+            {
+                if (_cookie != -1)
+                {
+                    _icp.Unadvise(_cookie);
+                }
+                if (_icp != null)
+                {
+                    Marshal.ReleaseComObject(_icp);
```
Comment (Vogel612) [inline Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs]: should we null out _icp here after this to keep Dispose and DetachEvents idempotent?

Comment (bclothier) [reply Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs]: Good catch! Will fix.

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 2]
```diff
@@ -113,22 +113,34 @@ public void RefreshDeclarationFinder()
...
+            _vbProjects.ProjectAdded -= Sinks_ProjectAdded;
+            _vbProjects.ProjectRemoved -= Sinks_ProjectRemoved;
+            _vbProjects.ProjectRenamed -= Sinks_ProjectRenamed;
+            foreach (var project in _vbProjects)
+            using (project)
+            {
+                project.VBComponents.ComponentAdded -= Sinks_ComponentAdded;
+                project.VBComponents.ComponentRemoved -= Sinks_ComponentRemoved;
+                project.VBComponents.ComponentRenamed -= Sinks_ComponentRenamed;
+                project.VBComponents.DetachEvents();
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: Please note that you are unsubscribing and detaching from on different components collection wrappers than you subscribed and attached to.

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 3]
```diff
@@ -113,22 +113,34 @@ public void RefreshDeclarationFinder()
...
+            _vbProjects.ProjectRemoved += Sinks_ProjectRemoved;
+            _vbProjects.ProjectRenamed += Sinks_ProjectRenamed;
+            foreach (var project in _vbProjects)
+            using (project)
+            {
+                project.VBComponents.AttachEvents();
+                project.VBComponents.ComponentAdded += Sinks_ComponentAdded;
+                project.VBComponents.ComponentRemoved += Sinks_ComponentRemoved;
+                project.VBComponents.ComponentRenamed += Sinks_ComponentRenamed;
+            }
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: This leaks components collection wrappers.

File: Rubberduck.VBEEditor/Events/VBENativeServices.cs [hunk 4]
```diff
@@ -63,8 +63,8 @@ public static void UnhookEvents()
                     info.Subclass.FocusChange -= FocusDispatcher;
                     info.Subclass.Dispose();
                 }
-                SafeComWrappers.VBA.VBProjects.DetatchEvents();
-                SafeComWrappers.VBA.VBComponents.DetatchEvents();
+                //SafeComWrappers.VBA.VBProjects.DetatchEvents();
+                //SafeComWrappers.VBA.VBComponents.DetatchEvents();
             }
```
Comment (MDoerner) [inline Rubberduck.VBEEditor/Events/VBENativeServices.cs]: This destroys the proper shutdown order: we want to detach all events before we start tearing down RD. (There is no guarantee in which order thisgs will get disposed on disposal of the IoC container. So, we do not want anything from the outside kick off anything while  we tear down.)

**Comment by bclothier** [comment]
This PR would be better served once #3705 is merged.

**Comment by bclothier** [comment]
This is still preliminary but I want to throw it out so others can confirm if the latest changes seem to be sane thing to do.

Based on @MDoerner 's comments, it seemed best that we make the repository the sole manager of the collection level events, so that we can reliably listen to instance-based events without losing track of the RCW. Going forward, all managed events should be attached to the project repository for the collections (IOW, at the IDE's level). The repository will then maintain the list of delegates so that when it has to refresh the components it can detach the delegates and re-attach, allowing the managed event subscribers to remain blissfully ignorant of COM management behind the scenes. 

TODO:

1) Update the refresh methods to handle detaching and attaching delegates
2) Actually test the code
3) Update other section of code to ensure that all collection-level events come from project repository and not the wrappers
4) Verify that we can use `DetachComponentsEvents` to support clean shutdown

File: Rubberduck.VBEEditor/ComManagement/ProjectsRepository.cs [hunk 5]
```diff
@@ -273,5 +289,124 @@ private void ClearCollections()
...
+                    project.VBComponents.ComponentAdded += (EventHandler<ComponentEventArgs>)subscriber;
+                    break;
+                case ComponentEventTypes.ComponentReloaded:
+                    project.VBComponents.ComponentReloaded += (EventHandler<ComponentEventArgs>)subscriber;
+                    break;
+                case ComponentEventTypes.ComponentRemoved:
+                    project.VBComponents.ComponentRemoved += (EventHandler<ComponentEventArgs>)subscriber;
+                    break;
+                case ComponentEventTypes.ComponentRenamed:
+                    project.VBComponents.ComponentRenamed += (EventHandler<ComponentRenamedEventArgs>)subscriber;
```
Comment (MDoerner) [inline Rubberduck.VBEEditor/ComManagement/ProjectsRepository.cs]: This leaks `VbComponents`.

**Comment by bclothier** [comment]
@MDoerner, thanks for the feedback and the leak catch. I will address that after below... I have few concerns I want to address before I work further on this.

From what I see, we really don't care (and don't think we should) which collection instance is raising the events at the collection levels. I'm more inclined to see it as IDE event, if you will, so I'd rather not burden the consumers with COM management for collection instances they have no control over.

1) Given that if the repository can refresh collections which includes destroying the instances and re-creating, which could also destroy the event subscriptions, doesn't it imply that the repository is responsible for restoring the existing delegates to the reloaded collections? 

2) Because repository is a singleton, and I can't really think of a good reason to not treat the collection events as originating from any other instances, we would be ending up with 2 singletons? We run into the same problem if we make events a part of VBE interface. 

3) For consumers that don't care about events, they are not required to listen to it, in which case, there will be no delegates added to the list to be maintained by the repository. 


If we still want to separate the event management and repository, I'm inclined to think that repository would have to have an event that is raised whenever the collection has been reloaded, as to notify the consuming client that they should refresh their collection instances. This does not feel clean.

Can anyone suggest a better way to architecture this? I'm all eyes.

**Comment by MDoerner** [comment]
I think it would be best to keep the event provider completely separate from the repository and to have its own collection of `VBComponents` collections, maybe in a dictionary with the `projectId` as key. The dictonary could get initializing using the `VBProjects` instance used for the project events, which itself could be acquired from an `IVBE` passed in through the constructor. Afterwards, the dictionary could be managed by reacting to the project added, project removed and project renamed events (before processing the registered handlers).

I see two advantages in this approach over combining the events provider with the `ProjectsRepository`. First, the `VBComponents` instances stay stable such that a reregistration is not necessary when a refresh of the `ProjectsRepository` happens. Second, the collections are always kept up to date.

I think the events provider and the `ProjectsRepository` should follow two different goals regarding their updating strategy. The events handler should always stay up-to-date in order not to miss component events. The `ProjectsRepository` on the other hand has an explicit `Refresh` method in order to allow the collections to stay stable while parsing; only on the next start of a parse, the method gets invoked and all disctionaries in the `ProjectsRepository` get refreshed.

**Comment by bclothier** [comment]
Just to confirm because I think you just called out a misconception I had - I was under the impression that when the VBComponents collection refreshes, the existing event subscriptions would no longer work because the underlying instance they reference would be disposed of and attached to a different instance. In prior review, you pointed out that if I do not hold a reference, I would leak events because they'd get attached/detached to a different instance. Thus, I must hold a reference to the collection instance, we can be assured that events won't leak. But if I take a reference from the `ProjectsRepository`, then I can't hold on my instance when the `ProjectsRepository` refreshes? 

Should I be _not_ taking any instances from `ProjectsRepository` but rather take them from VBE directly and re-wrapping them?

**Comment by MDoerner** [comment]
The lifetime and disposal of wrappers provided by the `ProjectsRepository` is managed by the repository. That is what most parts of RD need and it removes a lot of redundant COM requests. The wrappers on the other hand return new wrappers whose lifetime and diposal has to be managed by the caller. I think the latter is what the events provider actually needs. Moreover, since it does not expose the collections to its consumers, it is free to manage them however it likes to.

In short, I guess you should get the collections from the VBE, or more precisely, the `VBProjects` collection from the VBE and the `VBComponents` collections from the corresponding `VBProject` instances.

**Comment by bclothier** [comment]
Update: I did some tests and I can confirm that for Excel 2010, this PR will close #3650 now as we are now correctly picking up the components event from 2nd project.

I also tested the different case where a separate Excel instance is spawned. Events from that 2nd instance is not picked up (as expected) but the events from 1st instance continue to work.

This still suffers from the project not closing that should be corrected by PR #3764, however, so in those tests, the `ProjectRemoved` never ever fires.

Note that I provided an additional event, `EventsTerminated` as a means to warn all consumers that we are terminating events notification because there can be only one instance of `VBEEvents` running in a given appdomain, which means the first consumer to call `Terminate` will cut off other consumers' notifications.

**Comment by bclothier** [comment]
Note: Tested on Word, Excel and Access 2010. PowerPoint 2010 crashes at startup but the crash is beyond the scope of the PR.

File: Rubberduck.VBEEditor/Events/VbeEvents.cs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    public sealed class VBEEvents : IVBEEvents
+    {
+        private static VBEEvents _instance;
+        private static readonly object Lock = new object();
+        private readonly IVBProjects _projects;
+        private readonly Dictionary<string, IVBComponents> _components;
+
+        public static VBEEvents Initialize(IVBE vbe)
+        {
+            if (_instance == null)
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/Events/VbeEvents.cs]: the null-check should be made with the lock acquired.

File: Rubberduck.VBEEditor/Events/VbeEvents.cs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                    _instance = new VBEEvents(vbe);
+                }
+            }
+
+            return _instance;
+        }
+
+        public static void Terminate()
+        {
+            if (_instance == null)
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/Events/VbeEvents.cs]: same here: the lock should be acquired _before_ the null-check happens.

File: Rubberduck.VBEEditor/Events/VbeEvents.cs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+        public event EventHandler EventsTerminated;
+
+        #region IDisposable
+
+        private bool _disposed;
+
+        private void Dispose(bool disposing)
+        {
+            if (!_disposed && _projects != null)
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/Events/VbeEvents.cs]: should probably acquire a lock here, no?

File: Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+        protected override void Dispose(bool disposing)
+        {
+            DetachEvents();
+            base.Dispose(disposing);
+        }
+
+        public void AttachEvents()
+        {
+            if (IsWrappingNullReference)
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs]: I'd extend the lock scope up to here..

File: Rubberduck.VBEEditor/Events/VbeEvents.cs (hunk 8 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/Events/VbeEvents.cs]: No, this did start out implementing `IDisposable` but I later removed the `IDisposable`, made this method `private` so that `Terminate` (which already has a lock) is the *only* place that can dispose this singleton. So it's safe to assume there is already a lock present by time we enter this method. I suppose I might need to rename the method now that it's no longer a true disposing pattern.

Comment (retailcoder) [reply Rubberduck.VBEEditor/Events/VbeEvents.cs]: Nah, it's close enough to earn the name ;-)
I just somehow glossed over the `private` modifier

File: Rubberduck.VBEEditor/SafeComWrappers/VB6/VBProjects.cs [hunk 10]
```diff
@@ -8,7 +8,7 @@
 
 namespace Rubberduck.VBEditor.SafeComWrappers.VB6
 {
-    public class VBProjects : SafeComWrapper<VB.VBProjects>, IVBProjects
+    public class VBProjects : SafeEventedComWrapper<VB.VBProjects, VB._dispVBProjectsEvents>, IVBProjects
     {
         private static readonly Guid VBProjectsEventsGuid = new Guid("0002E190-0000-0000-C000-000000000046");
```
Comment (WaynePhillipsEA) [inline Rubberduck.VBEEditor/SafeComWrappers/VB6/VBProjects.cs]: same as above

File: Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+                // Call QueryInterface for IConnectionPointContainer
+                var icpc = (IConnectionPointContainer) Target;
+
+                // Find the connection point for the source interface
+                var g = typeof(TEventInterface).GUID;
+                icpc.FindConnectionPoint(ref g, out _icp);
+
+                // Pass a pointer to the host to the connection point
+                _icp.Advise(this as TEventInterface, out _cookie);
```
Comment (WaynePhillipsEA) [inline Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs]: I would prefer to see this as a proper cast to disallow the null passthrough which shouldn't be needed?

File: Rubberduck.VBEEditor/SafeComWrappers/VB6/VBComponents.cs [hunk 12]
```diff
@@ -11,7 +11,7 @@
 
 namespace Rubberduck.VBEditor.SafeComWrappers.VB6
 {
-    public class VBComponents : SafeComWrapper<VB.VBComponents>, IVBComponents
+    public class VBComponents : SafeEventedComWrapper<VB.VBComponents, VB._dispVBComponentsEvents>, IVBComponents
```
Comment (WaynePhillipsEA) [inline Rubberduck.VBEEditor/SafeComWrappers/VB6/VBComponents.cs]: I appreciate we're not supporting VB6 yet, but...  as we're not implementing the event interfaces for the VB6 versions here yet, I think it would be better to revert this change.  Anyway, I think the VB._dispVBProjectsEvents is not the correct interface in VB6 as it uses different, but very similar, source interfaces.    Maybe revert and add a TODO

File: Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs (hunk 11 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs]: Not sure I follow. The `SafeEventedComWrapper` would require the implementation to have `TEventInterface` defined and thus can be cast to that interface. Unfortunately, there is no semantic means in C# to express that a generic type should be implemented by the class so I cannot enforce that the implementing class _must_ implement the event interface but I can at least hint it should be since the interface is required by the `TEventInterface` argument of the abstract class.

Comment (WaynePhillipsEA) [reply Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs]: I mean by changing to (TEventInterface)this

**Comment by retailcoder** [comment]
PPT crash confirmed unrelated to this PR

File: Rubberduck.VBEEditor/SafeComWrappers/VB6/VBProjects.cs (hunk 10 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/SafeComWrappers/VB6/VBProjects.cs]: The problem is that we use the interface (`IVBComponents` and `IVBProjects`), which are shared by both VB6 and VBA, so if I don't implement VB6 using same interface, I get compilation errors because the consuming code can't call `AttachEvents` / `DetachEvents` since VB6's implementation won't have `SafeEventedComWrapper` interface. I agree it's lousy but short of actually deleting the whole VB6 implementation, I'm not sure what I can do about that. I will at least add a TODO to warn it's all wrong.

Comment (WaynePhillipsEA) [reply Rubberduck.VBEEditor/SafeComWrappers/VB6/VBProjects.cs]: Ah right, understood.  Yeah, a TODO will suffice.

**Comment by WaynePhillipsEA** [comment]
According to the IConnectionPoint::Advise documentation, https://msdn.microsoft.com/en-us/library/windows/desktop/ms678815(v=vs.85).aspx:

pdwCookie ... If the connection was not successfully established, this value is zero

Currently, the NotAdvising default const value for the cookie is set to -1, which is not correct.  It should be 0, as -1 could be a valid cookie value.

**Comment by bclothier** [comment]
Hmm. I'm a bit puzzled as to why sample (I should have saved the link) used `-1`. I thought probably to indicate that it was never advised before. 

Is there any reason for us to know whether we ever had advised? Otherwise I would simply change the const to 0 as per documentation.

**Comment by WaynePhillipsEA** [comment]
Only if you find a bad implementation of IConnectionPoint that returns a zero cookie as a valid value (against spec)

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 13]
```diff
@@ -349,7 +349,7 @@ public void SetModuleState(QualifiedModuleName module, ParserState state, Syntax
         private IVBProject GetProject(string projectId)
         {
             return _projectRepository.Project(projectId);
-        }
+                        }
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: Why the whitespace?

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 14]
```diff
@@ -952,6 +952,7 @@ public void Dispose()
 
             CoClasses?.Clear();
             RemoveEventHandlers();
+            VBEEvents.Terminate();
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: Why do we do this here in addition to doing it in `UnhookEvents`, which is calles in the shutdown code?

File: Rubberduck.VBEEditor/Events/VbeEvents.cs [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.VBEditor.SafeComWrappers;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+
+namespace Rubberduck.VBEditor.Events
+{
+    public sealed class VBEEvents : IVBEEvents
+    {
+        private static VBEEvents _instance;
```
Comment (MDoerner) [inline Rubberduck.VBEEditor/Events/VbeEvents.cs]: Why a static singleton instead of constructor injection and registration in the IoC container in singleton scope?

File: Rubberduck.VBEEditor/Events/VbeEvents.cs [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+            components.ComponentRemoved += ComponentRemovedHandler;
+            components.ComponentRenamed += ComponentRenamedHandler;
+            components.ComponentActivated += ComponentActivatedHandler;
+            components.ComponentSelected += ComponentSelectedHandler;
+            components.ComponentReloaded += ComponentReloadedHandler;
+        }
+
+        private void UnregisterComponents(string projectId)
+        {
+            if (!_components.ContainsKey(projectId))
```
Comment (MDoerner) [inline Rubberduck.VBEEditor/Events/VbeEvents.cs]: Why not `TryGet`? (It is on the `IDictionary` interface.)

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs (hunk 14 above)
Comment (bclothier) [reply Rubberduck.Parsing/VBA/RubberduckParserState.cs]: Because that was where the original code had terminated the events. The `UnhookEvents` merely deregister the class from listening to events but `Terminate` shutdowns *all* event notifications from all consumers. I decided to leave them in each place in case there was a need for the class to unhook events but without shutting down everyone else's notifications.

File: Rubberduck.VBEEditor/Events/VbeEvents.cs (hunk 15 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/Events/VbeEvents.cs]: Because the ctor is not parameterless and a normal singleton can only have parameterless static constructor. C# does not permit a static constructor to contain parameters. Furthermore, we need to be able to deterministically shut down all event notifications, which has to be done via a static method.

File: Rubberduck.VBEEditor/Events/VbeEvents.cs (hunk 16 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/Events/VbeEvents.cs]: Did not know about that. Will update.

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs (hunk 13 above)
Comment (bclothier) [reply Rubberduck.Parsing/VBA/RubberduckParserState.cs]: ... I'm not sure. I will remove.

File: Rubberduck.VBEEditor/Events/VbeEvents.cs (hunk 16 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/Events/VbeEvents.cs]: Hm. On further reflection, I have concerns about potential leaks. If you look a bit further, we have a `using(var components = _components[projectId])`. If I take a `VBComponents` as a `out` parameter, I'm not sure if I can safely manage that variable without leak?

File: Rubberduck.VBEEditor/Events/VbeEvents.cs (hunk 15 above)
Comment (MDoerner) [reply Rubberduck.VBEEditor/Events/VbeEvents.cs]: First, you can register as singleton in the IoC container with any constructor you want.

The second point is an actual argument not to let the IoC container deal with the disposal, but I do not think we need a static method for it. Instead, we could construct the event provider immediately before the IoC installer and let the IoC container inject that instance, just like we do with the VBE and AddIn instances. Then, we have access to it in the shutdown code and can terminate the instance there.

Well, I guess it also works with this static approach, but I have to say that I am not a fan of static code.

Comment (bclothier) [reply Rubberduck.VBEEditor/Events/VbeEvents.cs]: Thanks for mentioning the `RegisterConstantVbeAndAddIn` -- I can see more clearly how I could have done this using IoC which wasn't clear to me before. 

Reviewing the code again, I see that the only other place the `Terminate` would have been called was in the source code presenter which we already removed from the earlier PR (and even so it did not feel right to me). I will revise to allow use of IoC in similar manner to the VBE/AddIn so it becomes consistent. 

I only have one more concern. if I remove the private ctor and allow it to be constructed by others, what is there to prevent misuse of the class since multiple instance would destroy its ability to deterministically clean up events?
